lua/legendary/data/function.lua

```diff
@@ -2,8 +2,24 @@ local class = require('legendary.vendor.middleclass')
 local util = require('legendary.util')
 local Filters = require('legendary.data.filters')
 
+---Check if modes is an array of strings or itself a string
+---@param modes table
+---@return boolean
+local is_list_of_strings_or_string = function(modes)
+  if modes == nil or type(modes) == 'string' then
+    return true
+  end
+  for _, mode in ipairs(modes) do
+    if type(mode) ~= 'string' then
+      return false
+    end
+  end
+  return true
+end
+
 ---@class Function
 ---@field implementation function
+---@field mode_mappings string[]
```

Comment:
We shouldn't have the `mode_mappings` be a different type for functions than for everything else.

Take a look at how `Keymap` handles it. The type should be something similar to `ModeKeymap`, but you'll just need to modify it slightly to take different options (the options for a `Function`).

Reply:
Quite a lot for an overnight PR which I made without a prior knowledge of the project, isn't it?

I didn't intend to implement fully `keymap`-compatible mode-map, cuz:
* Nobody is going to drop funcs items into `legendary.keymap({ ... })`
* Even if they do, current modemap type is fully forward compatible with `legendary.keymap({ ... })`;
* Zero use cases



> _you need just slightly_

Only to make things more complicated: validation (see below) requires that item's `[1]`-field is provided with a function, so introduction of `keymaps`-like node-map will likely break something  - you have to figure you what to do with that function; it's subject of further discussion.

I suggest you take care of map enhancement further after merging this PR.

https://github.com/mrjones2014/legendary.nvim/blob/6ea46c285f3fd49317e04b6b31fad7c86b975cb0/lua/legendary/data/function.lua#L15-L20

Reply:
> Quite a lot for an overnight PR which I made without a prior knowledge of the project, isn't it?

Not sure what you mean. There's no deadline on the PR.

Regarding your 3 bullets, addressing in-order:

- That's not what I said and it's not what my comment is meant to support
- See above
- The use case is to support the alternate declaration syntax (2nd code block [here](https://github.com/mrjones2014/legendary.nvim/blob/master/doc/table_structures/KEYMAPS.md#specifying-mode))

> Only to make things more complicated: validation (see below) requires that item's [1]-field is provided with a function, so introduction of keymaps-like node-map will likely break something - you have to figure you what to do with that function; it's subject of further discussion.

It's not really more complicated. You can copy the validator from `keymap.lua` only for the options, validate that they're `Function` options instead of `Keymap` options. Then just update `executor.lua` to pull the right implementation out of the item based on the mode.

It we're going to introduce `mode_mappings` for `Function`s, then they're going to work the same as they do for `Keymap`s.

Reply:
> (2nd code block [here](https://github.com/mrjones2014/legendary.nvim/blob/master/doc/table_structures/KEYMAPS.md#specifying-mode))
>  then they're going to work the same as they do for Keymaps.

As I said, this PR is _about_ functions, not keymaps; if you want to unify mode-map apis for both, I would suggest you to take care of further mode-map improvement yourself taking the laid groundwork as basis; for me, it's more than enough to have `mode = { 'n', 'v', 'x' }` per one function; for more - i would just create more items with different `mode`; cheap and quick

And as I said, I see no reason to introduce that complicated approach, cause there are going to be [ambiguities](https://github.com/mrjones2014/legendary.nvim/pull/415#discussion_r1367045802) you will have to resolve as project owner